CMake: avrdude.conf dependency to avrdude.conf.in is broken

My workflow during development is like this: The first time I check out a branch I do
```
$ rm -rf build_linux
$ cmake -D CMAKE_BUILD_TYPE=RelWithDebInfo -B build_linux
$ cmake --build build_linux
```
After changes when I want to test, it's repeatedly the following, which nicely compiles and installs the to-be-tested avrdude:
```
cmake --build build_linux; sudo cmake --build build_linux --target install
```

However, I noticed that the systemwide `avrdude.conf` does not get installed/changed with the latter command even after explicitly changing `src/avrdude.conf.in`. It gets installed with the first-time creation of the build_linux dir.

What I'd like is a dependency of the systemwide conf file on the source conf.in. How do I effect that?

@dl8dtl 
BTW, even though `build.sh` only claims to support only Linux/macOS/FreeBSD, it works well under MYS2 mingw32/64. Of course the comments using `sudo` to install will not work under MSYS2 mingw32/64.

```
MINGW64 /c/work/avr/avrdude_test/avrdude_main
$ ./build.sh
-- Configuration summary:
-- ----------------------
-- DO HAVE    libelf
-- DO HAVE    libusb
-- DO HAVE    libusb_1_0
-- DO HAVE    libhidapi
-- DON'T HAVE libftdi
-- DO HAVE    libftdi1
-- DISABLED   doc
-- DISABLED   parport
-- DISABLED   linuxgpio
-- DISABLED   linuxspi
-- ----------------------
-- Configuring done
-- Generating done
-- Build files have been written to: C:/work/avr/avrdude_test/avrdude_main/build_mingw64_nt-10.0-22000
[38/56] Building C object src/CMakeFiles/libavrdude.dir/ser_win32.c.obj
C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c: In function 'net_send':
C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c:396:36: warning: pointer targets in passing argument 2 of 'send' differ in signedness [-Wpointer-sign]
  396 |                 rc = send(fd->ifd, p, (len > 1024) ? 1024 : len, 0);
      |                                    ^
      |                                    |
      |                                    const unsigned char *
In file included from C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c:30:
C:/msys64/mingw64/include/winsock2.h:1033:60: note: expected 'const char *' but argument is of type 'const unsigned char *'
 1033 |   WINSOCK_API_LINKAGE int WSAAPI send(SOCKET s,const char *buf,int len,int flags);
      |                                                ~~~~~~~~~~~~^~~
C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c: In function 'net_recv':
C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c:530:36: warning: pointer targets in passing argument 2 of 'recv' differ in signedness [-Wpointer-sign]
  530 |                 rc = recv(fd->ifd, p, (buflen - len > 1024) ? 1024 : buflen - len, 0);
      |                                    ^
      |                                    |
      |                                    unsigned char *
C:/msys64/mingw64/include/winsock2.h:1028:54: note: expected 'char *' but argument is of type 'unsigned char *'
 1028 |   WINSOCK_API_LINKAGE int WSAAPI recv(SOCKET s,char *buf,int len,int flags);
      |                                                ~~~~~~^~~
C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c: In function 'net_drain':
C:/work/avr/avrdude_test/avrdude_main/src/ser_win32.c:696:36: warning: pointer targets in passing argument 2 of 'recv' differ in signedness [-Wpointer-sign]
  696 |                 rc = recv(fd->ifd, &buf, 1, 0);
      |                                    ^~~~
      |                                    |
      |                                    unsigned char *
C:/msys64/mingw64/include/winsock2.h:1028:54: note: expected 'char *' but argument is of type 'unsigned char *'
 1028 |   WINSOCK_API_LINKAGE int WSAAPI recv(SOCKET s,char *buf,int len,int flags);
      |                                                ~~~~~~^~~
[53/56] Building C object src/CMakeFiles/avrdude.dir/term.c.obj
In function 'tokenize',
    inlined from 'terminal_mode' at C:/work/avr/avrdude_test/avrdude_main/src/term.c:1473:12:
C:/work/avr/avrdude_test/avrdude_main/src/term.c:1360:15: warning: pointer 'buf' may be used after 'realloc' [-Wuse-after-free]
 1360 |       k = buf - bufp;
      |           ~~~~^~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/term.c:1344:18: note: call to 'realloc' here
 1344 |       buf_tmp  = realloc(buf, bufsize);
      |                  ^~~~~~~~~~~~~~~~~~~~~
[54/56] Building C object src/CMakeFiles/libavrdude.dir/stk500v2.c.obj
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c: In function 'stk500v2_setparm':
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3098:6: warning: 'current_value' may be used uninitialized [-Wmaybe-uninitialized]
 3098 |   if (value == current_value) {
      |      ^
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3090:17: note: 'current_value' was declared here
 3090 |   unsigned char current_value;
      |                 ^~~~~~~~~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c: In function 'stk500v2_print_parms1':
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3295:24: warning: 'osc_cmatch' may be used uninitialized [-Wmaybe-uninitialized]
 3295 |       f /= (osc_cmatch + 1);
      |            ~~~~~~~~~~~~^~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3242:47: note: 'osc_cmatch' was declared here
 3242 |   unsigned char vtarget, vadjust, osc_pscale, osc_cmatch, sck_duration =0; //XXX 0 is not correct, check caller
      |                                               ^~~~~~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3242:35: warning: 'osc_pscale' may be used uninitialized [-Wmaybe-uninitialized]
 3242 |   unsigned char vtarget, vadjust, osc_pscale, osc_cmatch, sck_duration =0; //XXX 0 is not correct, check caller
      |                                   ^~~~~~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3278:5: warning: 'vadjust' may be used uninitialized [-Wmaybe-uninitialized]
 3278 |     avrdude_message(MSG_INFO, "%sVaref           : %.1f V\n", p, vadjust / 10.0);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3242:26: note: 'vadjust' was declared here
 3242 |   unsigned char vtarget, vadjust, osc_pscale, osc_cmatch, sck_duration =0; //XXX 0 is not correct, check caller
      |                          ^~~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3267:5: warning: 'vtarget' may be used uninitialized [-Wmaybe-uninitialized]
 3267 |     avrdude_message(MSG_INFO, "%sVtarget         : %.1f V\n", p, vtarget / 10.0);
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/work/avr/avrdude_test/avrdude_main/src/stk500v2.c:3242:17: note: 'vtarget' was declared here
 3242 |   unsigned char vtarget, vadjust, osc_pscale, osc_cmatch, sck_duration =0; //XXX 0 is not correct, check caller
      |                 ^~~~~~~
[56/56] Linking C executable src\avrdude.exe

Build succeeded.

Run

sudo cmake --build build_mingw64_nt-10.0-22000 --target install

to install.


xiaof@LAPTOP-913SIMUD MINGW64 /c/work/avr/avrdude_test/avrdude_main
$ cp ./build_mingw64_nt-10.0-22000/src/avrdude.exe ../avrdude_bin/avrdude
avrdude.conf                   avrdude.conf_v7.0              avrdude_git_main.exe           avrdude_pr1031.exe             avrdude_stk500v1_issue967.exe
avrdude.conf.new               avrdude_arduino_issue967.exe   avrdude_pr1030.exe             avrdude_pr1034.exe

xiaof@LAPTOP-913SIMUD MINGW64 /c/work/avr/avrdude_test/avrdude_main
$ cp ./build_mingw64_nt-10.0-22000/src/avrdude.exe ../avrdude_bin/avrdude_git.exe

xiaof@LAPTOP-913SIMUD MINGW64 /c/work/avr/avrdude_test/avrdude_main
$ ./build.sh
-- Configuration summary:
-- ----------------------
-- DO HAVE    libelf
-- DO HAVE    libusb
-- DO HAVE    libusb_1_0
-- DO HAVE    libhidapi
-- DON'T HAVE libftdi
-- DO HAVE    libftdi1
-- DISABLED   doc
-- DISABLED   parport
-- DISABLED   linuxgpio
-- DISABLED   linuxspi
-- ----------------------
-- Configuring done
-- Generating done
-- Build files have been written to: C:/work/avr/avrdude_test/avrdude_main/build_mingw64_nt-10.0-22000
ninja: no work to do.

Build succeeded.

Run

sudo cmake --build build_mingw64_nt-10.0-22000 --target install

to install.
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -282,27 +282,6 @@ endif()
 add_subdirectory(src)
 
 # =====================================
-# Setup default port names
-# =====================================
-
-if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
-    set(DEFAULT_PAR_PORT "/dev/parport0")
-    set(DEFAULT_SER_PORT "/dev/ttyS0")
-elseif (CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
-    set(DEFAULT_PAR_PORT "/dev/ppi0")
-    set(DEFAULT_SER_PORT "/dev/cuad0")
-elseif (CMAKE_SYSTEM_NAME STREQUAL "Solaris")
-    set(DEFAULT_PAR_PORT "/dev/printers/0")
-    set(DEFAULT_SER_PORT "/dev/term/a")
-elseif (CMAKE_SYSTEM_NAME STREQUAL "Windows")
-    set(DEFAULT_PAR_PORT "lpt1")
-    set(DEFAULT_SER_PORT "com1")
-else()
-    set(DEFAULT_PAR_PORT "unknown")
-    set(DEFAULT_SER_PORT "unknown")
-endif()
-
-# =====================================
 # Configuration
 # =====================================
 

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -82,29 +82,51 @@ else()
 endif()
 
 # =====================================
+# Setup default port names
+# =====================================
+
+if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
+    set(DEFAULT_PAR_PORT "/dev/parport0")
+    set(DEFAULT_SER_PORT "/dev/ttyS0")
+elseif (CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+    set(DEFAULT_PAR_PORT "/dev/ppi0")
+    set(DEFAULT_SER_PORT "/dev/cuad0")
+elseif (CMAKE_SYSTEM_NAME STREQUAL "Solaris")
+    set(DEFAULT_PAR_PORT "/dev/printers/0")
+    set(DEFAULT_SER_PORT "/dev/term/a")
+elseif (CMAKE_SYSTEM_NAME STREQUAL "Windows")
+    set(DEFAULT_PAR_PORT "lpt1")
+    set(DEFAULT_SER_PORT "com1")
+else()
+    set(DEFAULT_PAR_PORT "unknown")
+    set(DEFAULT_SER_PORT "unknown")
+endif()
+
+# =====================================
 # Configure files
 # =====================================
 
-macro(configure_option option)
-    if(${${option}})
-        string(REGEX REPLACE "(.*)@${option}_BEGIN@(.*)@${option}_END@(.*)" "\\1\\2\\3" conf_file "${conf_file}")
-    else()
-        string(REGEX REPLACE "(.*)@${option}_BEGIN@(.*)@${option}_END@(.*)" "\\1\\3" conf_file "${conf_file}")
-    endif()
-endmacro()
-
-file(READ avrdude.conf.in conf_file)
-configure_option(HAVE_PARPORT)
-configure_option(HAVE_LINUXGPIO)
-configure_option(HAVE_LINUXSPI)
-file(WRITE "${PROJECT_BINARY_DIR}/avrdude.conf.in" "${conf_file}")
-
 configure_file(cmake_config.h.in ac_cfg.h)
-configure_file("${PROJECT_BINARY_DIR}/avrdude.conf.in" avrdude.conf)
 configure_file(avrdude.spec.in avrdude.spec)
 if(WIN32)
     configure_file(windows.rc.in windows.rc)
 endif()
+
+add_custom_command(
+    OUTPUT avrdude.conf
+    COMMAND ${CMAKE_COMMAND} -E copy "${CMAKE_CURRENT_SOURCE_DIR}/avrdude.conf.in" avrdude.conf.in
+    COMMAND ${CMAKE_COMMAND}
+        -D HAVE_PARPORT=${HAVE_PARPORT}
+        -D HAVE_LINUXSPI=${HAVE_LINUXSPI}
+        -D HAVE_LINUXGPIO=${HAVE_LINUXGPIO}
+        -D DEFAULT_PAR_PORT=${DEFAULT_PAR_PORT}
+        -D DEFAULT_SER_PORT=${DEFAULT_SER_PORT}
+        -P "${CMAKE_CURRENT_SOURCE_DIR}/configure.cmake"
+    DEPENDS avrdude.conf.in
+    VERBATIM
+    )
+
+add_custom_target(conf ALL DEPENDS avrdude.conf)
 
 # =====================================
 # Project

--- a/src/configure.cmake
+++ b/src/configure.cmake
@@ -1,0 +1,35 @@
+#
+# configure.cmake - autoconf like multi-line configure
+# Copyright (C) 2022 Marius Greuel
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+# Do a multi-line replace based on @<OPTION>_BEGIN@ and @<OPTION>_END@ tags.
+macro(configure_option option)
+    if(${${option}})
+        string(REGEX REPLACE "(.*)@${option}_BEGIN@(.*)@${option}_END@(.*)" "\\1\\2\\3" CONTENTS "${CONTENTS}")
+    else()
+        string(REGEX REPLACE "(.*)@${option}_BEGIN@(.*)@${option}_END@(.*)" "\\1\\3" CONTENTS "${CONTENTS}")
+    endif()
+endmacro()
+
+# Perform autoconf like multi-line configure
+file(READ avrdude.conf.in CONTENTS)
+configure_option(HAVE_PARPORT)
+configure_option(HAVE_LINUXSPI)
+configure_option(HAVE_LINUXGPIO)
+file(WRITE avrdude.conf.in "${CONTENTS}")
+
+configure_file(avrdude.conf.in avrdude.conf)

